### DISCUSSION TITLE
Returning a value from an input widget.

### DISCUSSION THREAD

**Discussion by jtreagan** [fltk-rs/fltk-rs#1547] [General]
My goal is to return a usable value from an fltk.rs input widget.  To start with, I went to the [FLTK.rs book and copied the simple example provided there](https://fltk-rs.github.io/fltk-book/Input-&-Output.html) into a new function, made a few modifications, and ran it.  Worked fine.  Then I tried to have the function return the inputted value and nothing I've tried works.  I even let the AI that comes with my IDE give it a try and ended up with more mess.  Here's the latest non-working attempt:
```
use fltk::{prelude::*, *, window};

fn main() {
    let value = input_prompt("This is a prompt");
    println!("\n The inputted value is:  {:?}", value);
}

fn input_prompt(prompt: &str) -> String {
    let a = app::App::default();
    let mut win = window::Window::default().with_size(400, 300);
    win.make_resizable(true);
    let flex = group::Flex::default().with_size(100, 100).column().center_of_parent();
    let _prompttext = frame::Frame::default().with_label(prompt);
    let input = input::Input::default();
    let mut btn = button::Button::default().with_label("Submit");

    flex.end();
    win.end();
    win.show();

    btn.set_callback(move |_| {
        println!("\n In the function the value is:  {}", input.value());
        input.value()
    });
    a.run().unwrap();
}
```
The issue seems to center around how the inputted value is moved into the closure.  So, how do I get this working?

**Comment by MoAlyousef** [comment]
Almost there:
```rust
use fltk::{prelude::*, window, *};

fn main() {
    let value = input_prompt("This is a prompt");
    println!("\n The inputted value is:  {:?}", value);
}

fn input_prompt(prompt: &str) -> String {
    let a = app::App::default();
    let mut win = window::Window::default().with_size(400, 300);
    win.make_resizable(true);
    let flex = group::Flex::default()
        .with_size(100, 100)
        .column()
        .center_of_parent();
    let _prompttext = frame::Frame::default().with_label(prompt);
    let input = input::Input::default();
    let mut btn = button::Button::default().with_label("Submit");

    flex.end();
    win.end();
    win.show();

    btn.set_callback({
        let input = input.clone();
        move |_| {
            println!("\n In the function the value is:  {}", input.value());
        }
    });
    a.run().unwrap();
    input.value()
}
```
Widgets are reference types and can be cloned while referring to the same object. In the btn callback, we clone in a scope then pass the input to the callback closure.

**Reply by jtreagan** [reply]
Thanks @MoAlyousef.  That helps and the widget works fine.  The AI that goes with my IDE helped me better understand what was going on with this line:
```
        let input = input.clone();
```
although I'm still not real clear why that line is inside the button enclosure, rather than preceding it.

My next step is to eliminate the `Submit` button and have the inputted value return to the calling function when `Enter` is pressed.  I've been trying a few things, but haven't yet found anything that works.  Even that AI I mentioned hasn't been able to come up with a workable solution.  Any ideas?

**Reply by MoAlyousef** [reply]
You can try
```
input.set_trigger(CallbackTrigger::EnterKey); // or EnterKeyAlways
```

**Reply by jtreagan** [reply]
Thanks for the suggestion.  I had actually tried that already.  All it does is highlight the text in the input field and then waits for the `Submit` button to be pressed.  I'd give up and go with always having the `Submit` button there, but it just seems that with FLTK being so well developed over the years, there should be a way to accomplish this.  I'll keep messing around with it for awhile.  Maybe I'll find a good solution.

**Comment by MoAlyousef** [comment]
You would need to set the input's callback:
```rust
input.set_trigger(CallbackTrigger::EnterKey);
input.set_callback(move |input| {
    // submit info
});
```

**Reply by jtreagan** [reply]
That works good!  Thanks, @MoAlyousef.  Maybe someday I'll get the hang of using FLTK widgets.  Appreciate your help!
